Add support for Typst printing

xzdlj opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
As discussed in #24959 , a new printer(TypstPrinter) was implemented, which should mostly reuse the code of the LatexPrinter for convenience. 


#### TODO

- [x] Add
- [x] Mul
- [x] Pow
- [x] Sum
- [x] Log
- [x] Derivative
- [x] Permutation
- [x] Cross, Curl, Divergence
- [x] Limit
- [x] Function
- [x] Greek letters
- [x] Printing settings 

#### Brief description of what is fixed or changed
The TypstPrinter is mostly consistent with the LatexPrinter and tries to reuse its code.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* printing
  * Added TypstPrinter which supports the basic printing in Typst

<!-- END RELEASE NOTES -->

xzdlj committed:
TypstPrinter: add some basic printer methods

This is the first commit of TypstPrinter.
TypstPrinter supports Add, Mul, Float, Pow and Sum now.
The settings and greek letters are not available.

xzdlj committed:
TypstPrinter: add _print_log

As the same method in LatexPrinter, the `_print_log` only prints
 natural logarithm.

xzdlj committed:
TypstPrinter: add _print_Limit and _typst for Inf

"infinity" represents "oo" and "->" represents "\to"

xzdlj committed:
TypstPrinter: add _print_Symbol and translate

Notice:
1.Greek letters are mostly consistent with LaTex.
2.`epsilon` and `varepsilon` are special cases.
3.`hbar` is "plank.reduce", maybe there needs a other_symbol_dictionary
4.There is no direct way to express `avg`
5.No `mathscr`, `mathcal` as fallback

xzdlj committed:
change variable name to avoid confusion

replace "tex" with "typst"

xzdlj committed:
remove redundent import

xzdlj committed:
fix doctest for translate

xzdlj committed:
TypstPrinter: add some simple functions

Add _print_Expbase, _print_abs, _print_floor and _print_ceil.

Roseleaves commented:
https://github.com/OrangeX4/typst-sympy-calculator

I think this project above could just work. If only asked for his permission, sympy print to typst -- and even typst imported to sympy -- could just be completed.

xzdlj commented:
> https://github.com/OrangeX4/typst-sympy-calculator
> 
> I think this project above could just work. If only asked for his permission, sympy print to typst -- and even typst imported to sympy -- could just be completed.

Yes, It works well for the conversion between sympy and typst in most situations and can be a good reference.
Anyway, in my opinion, we still need a `TypstPrinter` for sympy as a subclass of `Printer` to handle more complex cases, such as the `unevaluate` option.

oscarbenjamin commented:
Looks like there is good work here although it is still marked as draft and the OP has some unticked checkboxes.

I haven't checked in detail whether there is anything significant missing. I guess maybe this should be documented somewhere?

xzdlj commented:
@oscarbenjamin Thanks.
The incomplete targets will be added soon, and I'll work on creating the related documentation page after that.

xzdlj committed:
TypstPrinter: add _print_Derivative

Some tests need `Integral` and small functions like `Max`.

xzdlj committed:
TypstPrinter: add _print_Function

xzdlj committed:
TypstPrinter: add permutation related functions

Repalce "\;" in latex with "space" and "pmatrix" with "mat".

xzdlj committed:
update the .mailmap to resolve conflict

xzdlj committed:
TypstPrinter: add some brackets help functions

Copy them from latex.py with nothing changed.

xzdlj committed:
TypstPrinter:add bool and notype related functions

xzdlj committed:
TypstPrinter: add vector related print functions

Modify Vector and Coordsys class.

xzdlj committed:
TypstPrinter: add logic functions

Represent <=/=> with Unicode which may not be proper.

xzdlj committed:
TypstPrinter: add matrix related functions

xzdlj committed:
TypstPrinter: Tensor and Array related functions

Note:Represent subscripts and superscripts with `scripts("")^` and
`scripts("")_` which is a useful workaround.

Also fix a little bug in _print_Mul.

xzdlj committed:
Add some missing variables and functions

xzdlj committed:
fix some subtle bugs and add new function

xzdlj committed:
add typst_repr, Prefix, Quantity, and units

- Add typst_repr to Prefix and Quantity classes.
- Implement _typst methods for Prefix, Quantity.

xzdlj committed:
finish most print functions
Need more work on tests and fix bugs.

xzdlj committed:
Implement Typst printer mostly

xzdlj commented:
Most of the implementation for the Typst printer has been completed, along with the corresponding tests. A code review would be appreciated. @oscarbenjamin 
Also, do I need to squash my commits since it's a bit messy.

oscarbenjamin commented:
I don't tend to do much with the printing code so I would rather hope that someone else reviews this than me.

What I will say though is that it seems unfortunate to have so much duplication. It seems like this ends up needing to duplicate almost the entirety of the code in the LaTeX printer including the methods needed for every type. The general structure of the printers would be a lot better if there was an intermediate display representation that could be shared between the printers rather than needing to duplicate code for the Cartesian product of printable objects foo and output printers bar.
